src/ProofTools/IterationTool.ts

```diff
@@ -0,0 +1,173 @@
+        if (currentNode instanceof CutNode) {
+            const color = isLegal(moveDifference, new Point(event.x - offset.x, event.y - offset.y))
+                ? legalColor()
+                : illegalColor();
+            drawAltered(currentNode, color, moveDifference);
+        } else if (currentNode instanceof AtomNode) {
+            const tempAtom: AtomNode = alterAtom(currentNode, moveDifference);
+            const color = isLegal(moveDifference, new Point(event.x - offset.x, event.y - offset.y))
+                ? legalColor()
+                : illegalColor();
+            drawAtom(tempAtom, color, true);
+        }
+    }
+}
+
+/**
+ * Checks to see if the node and any of its children are illegal in the new position. If any are
+ * illegal does not insert any of them. Redraws the canvas to update this.
+ * @param event The mouse out event while using the iteration tool
+ */
+export function iterationMouseUp(event: MouseEvent) {
+    if (legalNode) {
+        const moveDifference: Point = new Point(
+            event.x - startingPoint.x,
+            event.y - startingPoint.y
+        );
+
+        if (currentNode instanceof CutNode) {
+            if (isLegal(moveDifference, new Point(event.x - offset.x, event.y - offset.y))) {
+                insertChildren(currentNode, moveDifference);
+            }
+        } else if (currentNode instanceof AtomNode) {
+            const tempAtom: AtomNode = alterAtom(currentNode, moveDifference);
+
+            if (isLegal(moveDifference, new Point(event.x - offset.x, event.y - offset.y))) {
+                treeContext.tree.insert(tempAtom);
+            }
+        }
+    }
+    redrawTree(treeContext.tree);
+    legalNode = false;
+}
+
+/**
+ * If the mouse has left the canvas then assume it is now illegal and reset the tree.
+ */
+export function iterationMouseOut() {
+    legalNode = false;
+    redrawTree(treeContext.tree);
+}
+
+/**
+ * If the current point is legally iterated (within the parent of the original node)
+ * checks if the altered nodes can be inserted, if all of them can returns true.
+ * @param moveDifference The change from the currentNodes original position
+ * @param currentPoint The current point on the graph the mouse is
+ * @returns Whether or not this location is legal
+ */
+function isLegal(moveDifference: Point, currentPoint: Point): boolean {
+    if (currentParent !== null && currentParent.containsPoint(currentPoint)) {
+        if (
+            currentNode instanceof CutNode &&
+            validateChildren(currentNode, moveDifference) &&
+            insertChildless(treeContext.tree.sheet, alterCut(currentNode, moveDifference).ellipse!)
+        ) {
+            return true;
+        } else if (
+            currentNode instanceof AtomNode &&
+            treeContext.tree.canInsert(alterAtom(currentNode, moveDifference))
+        ) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    return false;
+}

```

Comment:
Ok finally getting to all of these, when you say single line return condition do you mean the entire function can be 1 line? Just want to make sure because my brain has been scrambled after work.

Reply:
Yes